File: core/src/test/scala/diplomacy/variant/standard/rule/BuildValidatorSpec.scala
package diplomacy.variant.standard.rule

import diplomacy.UnitSpec
import diplomacy.board._
import diplomacy.rule.InvalidOrderMessage
import diplomacy.variant.standard.map
import diplomacy.variant.standard.map._
import diplomacy.variant.standard.board.{ Turn => T }
import diplomacy.variant.standard.board._

class BuildVailidatorSpec extends UnitSpec {
  val validator = new BuildValidator[T, map.Power]
  val board = {
    Board[State[T], map.Power, MilitaryBranch.MilitaryBranch, UnitStatus.UnitStatus, ProvinceStatus.ProvinceStatus](
      map.map,
      1901.Autumn - Build,
      Set(
        DiplomacyUnit(Germany, Army, Mun),
        DiplomacyUnit(Italy, Fleet, Nap)
      ),
      Map(
        Rom.province -> Italy, Nap.province -> Italy,
        Mar.province -> France
      ),
      Map(), Map()
    )
  }
  val $ = StandardRuleOrderHelper(board)
  import $._

  "A build-validator" when {
    "a valid order is received" should {
      "return null (1)." in {
        validator.errorMessageOfOrder(board)(Rom.build(Army)) should be(None)
      }
      "return null (2)." in {
        validator.errorMessageOfOrder(board)(A(Mun).disband()) should be(None)
      }
    }

    "try to build an unit to a location that has another unit" should {
      "return an error message." in {
        validator.errorMessageOfOrder(board)(Nap.build(Army)) should be(Some(
          InvalidOrderMessage("An unit is in Nap.")
        ))
      }
    }
    "try to build an unit to a province that is not home province" should {
      "return an error message." in {
        validator.errorMessageOfOrder(board)(Order.Build(DiplomacyUnit(France, Army, Spa))) should be(Some(
          InvalidOrderMessage("France cannot build an unit in Spa.")
        ))
      }
    }
    "try to build an unit to a province that is not supply center" should {
      "return an error message." in {
        validator.errorMessageOfOrder(board)(Ruh.build(Army)) should be(Some(
          InvalidOrderMessage("Ruh is not supply center.")
        ))
      }
    }
    "try to build an unit to a province that is not occupied" should {
      "return an error message." in {
        validator.errorMessageOfOrder(board)(Bre.build(Army)) should be(Some(
          InvalidOrderMessage("Bre is not occupied by France.")
        ))
      }
    }

    "try to disband an unit that does not exist" should {
      "return an error message." in {
        validator.errorMessageOfOrder(board)(DiplomacyUnit(Germany, Army, Ber).disband()) should be(Some(
          InvalidOrderMessage("A Ber does not exist.")
        ))
      }
    }
    "try an unnecessary disband order" should {
      "return an error message." in {
        validator.errorMessageOfOrder(board)(F(Nap).disband()) should be(Some(
          InvalidOrderMessage("Italy has sufficient supply centers.")
        ))
      }
    }

    "valid order set is received" should {
      "return null." in {
        validator.errorMessageOfOrders(board)(Set(A(Mun).disband())) should be(None)
      }
    }
    "try to keep units that are more than supply centers" should {
      "return an error message." in {
        validator.errorMessageOfOrders(board)(Set()) should be(Some(
          InvalidOrderMessage("Germany does not have enough supply centers.")
        ))
      }
    }
  }
}

File: test/variant/standard/rule/build-validator-spec.ts
import * as chai from "chai"
import {
  BuildValidator, Board, Unit, MilitaryBranch, State, Result, Phase, ProvinceStatus,
  StandardRuleHelper as Helper
} from "../../../../src/variant/standard/rule/module"
import * as Order from "./../../../../src/variant/standard/rule/order"
import * as Error from "./../../../../src/variant/standard/rule/error"
import { Executed } from "./../../../../src/rule/module"
import { locations as $, Power, map } from "./../../../../src/variant/standard/map/module"
import { Season, Turn } from "./../../../../src/variant/standard/board/module"

const { Army, Fleet } = MilitaryBranch
const { Build } = Phase

const should = chai.should()

const validator = new BuildValidator()

describe("A BuildValidator", () => {
  const board = new Board(
    map, new State(new Turn(1901, Season.Autumn), Build),
    [
      new Unit(Army, $.Mun, Power.Germany),
      new Unit(Fleet, $.Nap, Power.Italy)
    ],
    [],
    [
      [$.Rom.province, new ProvinceStatus(Power.Italy, false)],
      [$.Nap.province, new ProvinceStatus(Power.Italy, false)],
      [$.Mar.province, new ProvinceStatus(Power.France, false)]
    ]
  )
  const $$ = new Helper(board)
  describe("when a valid order is received", () => {
    it("returns null (1)", () => {
      should.equal(validator.errorMessageOfOrder(board, $$.A($.Rom).build()), null)
    })
    it("returns null (2)", () => {
      should.equal(validator.errorMessageOfOrder(board, $$.A($.Mun).disband()), null)
    })
  })
  describe("when try to build an unit to a location that has another unit", () => {
    it("returns an error message", () => {
      new Error.UnbuildableLocation($$.A($.Nap).unit).should.deep.equal(
        validator.errorMessageOfOrder(board, $$.A($.Nap).build())
      )
    })
  })
  describe("when try to build an unit to a province that is not home province", () => {
    it("returns an error message", () => {
      new Error.UnbuildableLocation(new Unit(Army, $.Spa, Power.France)).should.deep.equal(
        validator.errorMessageOfOrder(
          board, new Order.Build(new Unit(Army, $.Spa, Power.France))
        )
      )
    })
  })
  describe("when try to build an unit to a province that is not supply center", () => {
    it("returns an error message", () => {
      new Error.UnbuildableLocation($$.A($.Ruh).unit).should.deep.equal(
        validator.errorMessageOfOrder(board, $$.A($.Ruh).build())
      )
    })
  })
  describe("when try to build an unit to a province that is not occupied", () => {
    it("returns an error message", () => {
      new Error.UnbuildableLocation($$.A($.Bre).unit).should.deep.equal(
        validator.errorMessageOfOrder(board, $$.A($.Bre).build())
      )
    })
  })
  describe("when try to disband an unit that does not exist", () => {
    it("returns an error message", () => {
      new Error.UnitNotExisted(new Unit(Army, $.Ber, Power.Germany)).should.deep.equal(
        validator.errorMessageOfOrder(
          board, new Order.Disband(new Unit(Army, $.Ber, Power.Germany))
        )
      )
    })
  })
  describe("when try to an unnecessary disband order", () => {
    it("returns an error message", () => {
      new Error.PowerWithProblem(Power.Italy).should.deep.equal(
        validator.errorMessageOfOrder(board, $$.F($.Nap).disband())
      )
    })
  })
  describe("when valid order set is received", () => {
    it("returns null", () => {
      should.equal(
        validator.errorMessageOfOrders(board, new Set([$$.A($.Mun).disband()])),
        null
      )
    })
  })
  describe("when try to keep units that are more than supply centers", () => {
    it("returns an error message", () => {
      new Error.PowerWithProblem(Power.Germany).should.deep.equal(
        validator.errorMessageOfOrders(board, new Set())
      )
    })
  })
})
